### PR TITLE
fix: cache issue

### LINKED ISSUE #1841: test: add test for cache-location flag from webpack@5
**Describe the bug**

supply location to cache path

**What is the current behavior?**
no tests present

**Expected behavior**
- tests present

**Screenshots**

**Please paste the results of `webpack-cli info` here, and mention other relevant information**

**Additional context**
/cc @evilebottnawi 
https://github.com/webpack/webpack-cli/pull/1832#discussion_r495111318

### LINKED ISSUE #1842: Project with `webpack-cli@4.0.0-rc.0` doesn't use cache
**Describe the bug**

I seem to have found a regression in webpack-cli 4 with regards to Webpack 5 caching.

Pretty minimal demo project: https://github.com/MarcelGerber/webpack-demo

With `webpack-cli@4.0.0-rc.0` and `webpack@5.0.0-rc.0`, the cache is not used even though it is specified in `webpack.config.js`. On [branch `old-cli`](https://github.com/MarcelGerber/webpack-demo/tree/old-cli) the cache is used, even though [the only change there](https://github.com/MarcelGerber/webpack-demo/commit/0a4150558be1651f631f9ccca13465c4111dcc04) is downgrading to `webpack-cli@3.3.12`.

**What is the current behavior?**

With `webpack-cli@4.0.0-rc.0` and `infrastructureLogging.debug: /webpack\.cache/` enabled, the output of `yarn webpack` is the following:
```
    [webpack.cache.PackFileCacheStrategy] No pack exists at /mnt/c/Users/marce/Programming/webpack-demo/node_modules/.cache/webpack/default-production.pack: Error: ENOENT: no such file or directory, open '/mnt/c/Users/marce/Programming/webpack-demo/node_modules/.cache/webpack/default-production/index.pack'
<t> [webpack.cache.PackFileCacheStrategy] restore cache container: 71.8964 ms
asset app.bundle.js 46.2 KiB [compared for emit] [minimized] (name: app)
asset print.bundle.js 862 bytes [compared for emit] [minimized] (name: print)
runtime modules 5.8 KiB 9 modules
cacheable modules 11.3 KiB
  modules by path ./src/ 2.16 KiB
    modules by path ./src/*.js 1.19 KiB
      ./src/main.js 1.03 KiB [built] [code generated]
      ./src/print.js 164 bytes [built] [code generated]
    modules by path ./src/*.css 996 bytes
      ./src/styles.css 520 bytes [built] [code generated]
      ./node_modules/css-loader/dist/cjs.js!./src/styles.css 476 bytes [built] [code generated]
  modules by path ./node_modules/ 9.1 KiB
    ./node_modules/style-loader/dist/runtime/injectStylesIntoStyleTag.js 6.64 KiB [built] [code generated]
    ./node_modules/css-loader/dist/runtime/api.js 2.46 KiB [built] [code generated]
webpack 5.0.0-rc.0 compiled successfully in 2948 ms
Done in 6.95s.
```

Notice how it only attempts to restore the cache, but never writes to it.

**To Reproduce**

1. Clone https://github.com/MarcelGerber/webpack-demo
2. Run `yarn`
3. Run `yarn webpack`
4. => You'll see that a `node_modules/.cache` folder is not created.
5. `git checkout old-cli && yarn`
6. `yarn webpack`
7. => Cache folder is created and caching works.

**Expected behavior**

With `webpack-cli@3.3.12`, the output when first running `yarn webpack` (without a pre-existing cache) is this:
```
    [webpack.cache.PackFileCacheStrategy] No pack exists at /mnt/c/Users/marce/Programming/webpack-demo/node_modules/.cache/webpack/default-production.pack: Error: ENOENT: no such file or directory, open '/mnt/c/Users/marce/Programming/webpack-demo/node_modules/.cache/webpack/default-production/index.pack'
<t> [webpack.cache.PackFileCacheStrategy] restore cache container: 72.8681 ms
    [webpack.cache.PackFileCacheStrategy] Pack got invalid because of write to: ResolverCachePlugin|normal|dependencyType=|esm|path=|/mnt/c/Users/marce/Programming/webpack-demo|request=|./src/main.js
    [webpack.cache.PackFileCacheStrategy] Storing pack...
    [webpack.cache.PackFileCacheStrategy] Capturing build dependencies... (/mnt/c/Users/marce/Programming/webpack-demo/webpack.config.js, /mnt/c/Users/marce/Programming/webpack-demo/node_modules/webpack/lib/, /mnt/c/Users/marce/Programming/webpack-demo/node_modules/style-loader/dist/cjs.js, /mnt/c/Users/marce/Programming/webpack-demo/node_modules/css-loader/dist/cjs.js)
<t> [webpack.cache.PackFileCacheStrategy] resolve build dependencies: 2253.9068 ms
<t> [webpack.cache.PackFileCacheStrategy] snapshot build dependencies: 120.1707 ms
    [webpack.cache.PackFileCacheStrategy] Captured build dependencies
    [webpack.cache.PackFileCacheStrategy] 33 fresh items in cache
<t> [webpack.cache.PackFileCacheStrategy] store pack: 227.7181 ms
    [webpack.cache.PackFileCacheStrategy] Stored pack
asset app.bundle.js 46.2 KiB [compared for emit] [minimized] (name: app)
asset print.bundle.js 862 bytes [compared for emit] [minimized] (name: print)
runtime modules 5.8 KiB 9 modules
cacheable modules 11.3 KiB
  modules by path ./src/ 2.16 KiB
    modules by path ./src/*.js 1.19 KiB
      ./src/main.js 1.03 KiB [built] [code generated]
      ./src/print.js 164 bytes [built] [code generated]
    modules by path ./src/*.css 996 bytes
      ./src/styles.css 520 bytes [built] [code generated]
      ./node_modules/css-loader/dist/cjs.js!./src/styles.css 476 bytes [built] [code generated]
  modules by path ./node_modules/ 9.1 KiB
    ./node_modules/style-loader/dist/runtime/injectStylesIntoStyleTag.js 6.64 KiB [built] [code generated]
    ./node_modules/css-loader/dist/runtime/api.js 2.46 KiB [built] [code generated]

WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value. Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/configuration/mode/

webpack 5.0.0-rc.0 compiled with 1 warning in 3273 ms
Done in 9.13s.
```

Notice that there's way more cache-related lines in there, and the cache is actually stored.

**Screenshots**

**Please paste the results of `webpack-cli info` here, and mention other relevant information**

```
  System:
    OS: Linux 4.4 Ubuntu 20.04.1 LTS (Focal Fossa), running inside WSL1 (Windows Subsystem for Linux)
    CPU: (8) x64 Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz
    Memory: 6.79 GB / 15.75 GB
  Binaries:
    Node: 12.13.1 - /tmp/yarn--1601152238051-0.21969676372665314/node
    Yarn: 1.22.5 - /tmp/yarn--1601152238051-0.21969676372665314/yarn
    npm: 6.14.8 - ~/.nvm/versions/node/v12.13.1/bin/npm
  Packages:
    webpack: 5.0.0-rc.0 => 5.0.0-rc.0
    webpack-cli: 4.0.0-rc.0 => 4.0.0-rc.0
    webpack-dev-server: ^3.11.0 => 3.11.0

Done in 0.97s.
```

**Additional context**

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**
Yes
**If relevant, did you update the documentation?**
No
**Summary**
Fix caching issue.
Fixes #1841 
Fixes #1842 

**Does this PR introduce a breaking change?**

**Other information**
No

### CODE REVIEW

File: packages/webpack-cli/lib/utils/Compiler.js [hunk 1]
```diff
@@ -87,8 +87,15 @@ class Compiler {
         // eslint-disable-next-line  no-async-promise-executor
         return new Promise(async (resolve) => {
             await this.compiler.run((err, stats) => {
-                const content = this.compilerCallback(err, stats, lastHash, options, outputOptions);
-                resolve(content);
+                if (this.compiler.close) {
```
Comment (anshumanv) [inline packages/webpack-cli/lib/utils/Compiler.js:90]: are there instances when `.close` is not available?

Comment (rishabh3112) [reply packages/webpack-cli/lib/utils/Compiler.js:90]: On versions other than v5

**Review by rishabh3112 — APPROVED** [review]

**Review by anshumanv — APPROVED** [review]

**Review by evenstensberg — APPROVED** [review]

File: test/cache/cache.test.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+const { run, isWebpack5 } = require('../utils/test-utils');
+
+describe('cache related tests', () => {
+    it('should log warning in case of single compiler', () => {
+        let { stderr, stdout } = run(__dirname, ['-c', 'webpack.config.js'], false);
+        // run 2nd compilation
+        ({ stderr, stdout } = run(__dirname, ['-c', 'webpack.config.js'], false));
```
Comment (evilebottnawi) [inline test/cache/cache.test.js:9]: I think we need use `async run` here

**Review by evilebottnawi — APPROVED** [review]

**Pull request merged**